File: backend/app/schemas/portal.py
import uuid
from datetime import datetime, timedelta
from decimal import Decimal

from pydantic import BaseModel

from app.models.escrow import Escrow, EscrowState
from app.schemas.bid import ParticipantOut
from app.schemas.money import Money

# Escrow releases to the seller two months after the sale (client rule).
PAYOUT_DAYS = 60


class BuyerDashboardOut(BaseModel):
    active_bids: int
    watchlist: int
    won_auctions: int
    purchases: int


class SellerDashboardOut(BaseModel):
    total_listings: int
    active_auctions: int
    total_earnings: Decimal
    pending_payouts: Decimal


class PurchaseOut(BaseModel):
    """A buyer's completed purchase, tracked through its escrow."""

    escrow_id: uuid.UUID
    property_id: uuid.UUID
    property_title: str
    property_image_url: str | None
    amount: Decimal
    state: EscrowState
    delivered_at: datetime | None
    purchased_at: datetime

    @classmethod
    def of(cls, escrow: Escrow) -> "PurchaseOut":
        return cls(
            escrow_id=escrow.id,
            property_id=escrow.property_id,
            property_title=escrow.listing.title,
            property_image_url=escrow.listing.image_url,
            amount=escrow.amount,
            state=escrow.state,
            delivered_at=escrow.delivered_at,
            purchased_at=escrow.created_at,
        )


class SellerEscrowOut(BaseModel):
    """A seller's escrow with its payout status and release countdown."""

    escrow_id: uuid.UUID
    property_id: uuid.UUID
    property_title: str
    buyer_id: uuid.UUID | None
    amount: Decimal
    state: EscrowState
    delivered_at: datetime | None
    created_at: datetime
    release_eta: datetime

    @classmethod
    def of(cls, escrow: Escrow) -> "SellerEscrowOut":
        return cls(
            escrow_id=escrow.id,
            property_id=escrow.property_id,
            property_title=escrow.listing.title,
            buyer_id=escrow.buyer_id,
            amount=escrow.amount,
            state=escrow.state,
            delivered_at=escrow.delivered_at,
            created_at=escrow.created_at,
            release_eta=escrow.created_at + timedelta(days=PAYOUT_DAYS),
        )


class ReservePriceUpdate(BaseModel):
    reserve_price: Money


class AuctionAnalysisOut(BaseModel):
    auction_id: uuid.UUID
    title: str
    reserve_price: Decimal
    winning_bid: Decimal | None
    reserve_met: bool
    total_bids: int
    bidder_count: int
    participants: list[ParticipantOut]


class DocumentVerificationOut(BaseModel):
    kyc_pending: int
    kyc_approved: int
    kyc_rejected: int
    escrow_open: int
    escrow_released: int

File: frontend/types/portal.ts
// export interface BuyerDashboard {
//   active_bids: number;
//   watchlist: number;
//   won_auctions: number;
//   purchases: number;
// }

// export interface SellerDashboard {
//   total_listings: number;
//   active_auctions: number;
//   total_earnings: string;
//   pending_payouts: string;
// }

// export type EscrowState = "awaiting_payment" | "held" | "released";

// export interface Purchase {
//   escrow_id: string;
//   property_id: string;
//   property_title: string;
//   property_image_url: string | null;
//   amount: string;
//   state: EscrowState;
//   delivered_at: string | null;
//   purchased_at: string;
// }

// export interface SellerEscrow {
//   escrow_id: string;
//   property_id: string;
//   property_title: string;
//   buyer_id: string | null;
//   amount: string;
//   state: EscrowState;
//   delivered_at: string | null;
//   created_at: string;
//   release_eta: string;
// }

// export interface AuctionAnalysis {
//   auction_id: string;
//   title: string;
//   reserve_price: string;
//   winning_bid: string | null;
//   reserve_met: boolean;
//   total_bids: number;
//   bidder_count: number;
//   participants: Array<{ user_id: string; full_name: string; bid_count: number }>;
// }

// export interface DocumentVerification {
//   kyc_pending: number;
//   kyc_approved: number;
//   kyc_rejected: number;
//   escrow_open: number;
//   escrow_released: number;
// }

export interface BuyerDashboard {
  active_bids: number;
  watchlist: number;
  won_auctions: number;
  purchases: number;
}

export interface SellerDashboard {
  total_listings: number;
  active_auctions: number;
  total_earnings: string;
  pending_payouts: string;
}

export type EscrowState = "awaiting_payment" | "held" | "released";

export interface Purchase {
  escrow_id: string;
  property_id: string;
  property_title: string;
  property_image_url: string | null;
  amount: string;
  state: EscrowState;
  delivered_at: string | null;
  purchased_at: string;
}

export interface SellerEscrow {
  escrow_id: string;
  property_id: string;
  property_title: string;
  buyer_id: string | null;
  amount: string;
  state: EscrowState;
  delivered_at: string | null;
  created_at: string;
  release_eta: string;
}

export interface AuctionAnalysis {
  auction_id: string;
  title: string;
  reserve_price: string;
  winning_bid: string | null;
  reserve_met: boolean;
  total_bids: number;
  bidder_count: number;
  participants: Array<{ user_id: string; full_name: string; bid_count: number }>;
}

export type BankAccountType = "savings" | "current";

export interface BankAccount {
  id: string;
  account_holder_name: string;
  bank_name: string;
  account_number_masked: string;
  ifsc_code: string;
  branch_name: string | null;
  account_type: BankAccountType;
  is_verified: boolean;
  updated_at: string;
}

export interface UpsertBankAccountRequest {
  account_holder_name: string;
  bank_name: string;
  account_number: string;
  ifsc_code: string;
  branch_name?: string | null;
  account_type: BankAccountType;
}

export interface DocumentVerification {
  kyc_pending: number;
  kyc_approved: number;
  kyc_rejected: number;
  escrow_open: number;
  escrow_released: number;
}
